app/models/user.rb

```diff
@@ -75,6 +77,17 @@ def self.reset_token_expired?(sent_at)
     Time.current > (sent_at.in(RESET_TOKEN_VALIDITY_PERIOD))
   end
 
+  def generate_session_token!(extended_session: false)
+    digest = User.generate_digest(SecureRandom.alphanumeric(40))
+    expiry = extended_session ? 7.days.from_now : 6.hours.from_now
+
+    update! session_token: digest, session_expiry: expiry
+  rescue ActiveRecord::RecordInvalid
+    raise unless errors.attribute_names.include? :session_token
+
+    retry
+  end
```

Comment:
Without the model level validation, I think this will not give the expected outcome.